File: control-plane/internal/cli/catalog.go
package cli

import (
	"fmt"
	"io"
	"os"

	"github.com/Agent-Field/agentfield/control-plane/internal/ui"
	"github.com/spf13/cobra"
)

// nodeCatalogEntry describes one installable AgentField agent node.
//
// This is deliberately a hand-vetted, in-binary list so `af catalog` works
// offline and gives a harness a curated set of nodes to install before any
// registry search lands. It is seeded from the desktop app's curated list
// (desktop/src/shared/catalog.ts) — keep the two in sync when adding nodes.
// `name` MUST equal the node's agentfield-package.yaml `name:` (the registry
// key after install), which is often not the repo name (SWE-AF → swe-planner).
type nodeCatalogEntry struct {
	Name        string `json:"name"`
	Description string `json:"description"`
	Source      string `json:"source"` // pass to `af install <source>`
	Docs        string `json:"docs"`   // public repo/docs URL
	Language    string `json:"language"`
}

// nodeCatalog is the curated set of installable nodes. Mirrors
// desktop/src/shared/catalog.ts. Docs point at the public GitHub repo (the
// `//<subdir>` source selector stripped, since the docs live at the repo root).
var nodeCatalog = []nodeCatalogEntry{
	{
		Name:        "swe-planner",
		Description: "Autonomous software-engineering fleet: plan, code, test, and ship production-grade PRs",
		Source:      "https://github.com/Agent-Field/SWE-AF",
		Docs:        "https://github.com/Agent-Field/SWE-AF",
		Language:    "python",
	},
	{
		Name:        "swe-planner-go",
		Description: "Go port of the SWE fleet: same planning/execution reasoners, one static binary",
		Source:      "https://github.com/Agent-Field/SWE-AF//go",
		Docs:        "https://github.com/Agent-Field/SWE-AF",
		Language:    "go",
	},
	{
		Name:        "pr-af",
		Description: "Turns a plain task description into a draft pull request on GitHub",
		Source:      "https://github.com/Agent-Field/pr-af",
		Docs:        "https://github.com/Agent-Field/pr-af",
		Language:    "python",
	},
	{
		Name:        "pr-af-go",
		Description: "Go port of the PR review agent: same reasoners, one static binary",
		Source:      "https://github.com/Agent-Field/pr-af//go",
		Docs:        "https://github.com/Agent-Field/pr-af",
		Language:    "go",
	},
	{
		Name:        "sec-af",
		Description: "Code security auditor: scans repositories and proves exploitability with verdicts and traces",
		Source:      "https://github.com/Agent-Field/sec-af",
		Docs:        "https://github.com/Agent-Field/sec-af",
		Language:    "python",
	},
	{
		Name:        "cloudsecurity-af",
		Description: "Cloud security posture: read-only attack-path scans across AWS, GCP, and Azure",
		Source:      "https://github.com/Agent-Field/cloudsecurity-af",
		Docs:        "https://github.com/Agent-Field/cloudsecurity-af",
		Language:    "python",
	},
}

// NewCatalogCommand builds `af catalog` — the browse step of the golden path,
// letting a harness discover installable nodes before `af install`.
func NewCatalogCommand() *cobra.Command {
	var outputFormat string
	cmd := &cobra.Command{
		Use:   "catalog",
		Short: "Browse installable AgentField agent nodes",
		Args:  cobra.NoArgs,
		RunE: func(cmd *cobra.Command, args []string) error {
			return runCatalog(os.Stdout, autoOutputFormat(outputFormat, isOutputTerminal()))
		},
		SilenceUsage: true,
	}
	cmd.Flags().StringVarP(&outputFormat, "output", "o", "", "Output format: pretty, json, yaml")
	return cmd
}

func runCatalog(stdout io.Writer, format string) error {
	switch format {
	case "json", "yaml":
		return writeValue(stdout, nodeCatalog, format)
	case "pretty":
		renderNodeCatalog(stdout)
		return nil
	default:
		return cliExitError{Code: 2, Err: fmt.Errorf("output format must be pretty, json, or yaml")}
	}
}

func renderNodeCatalog(stdout io.Writer) {
	rows := make([][]string, 0, len(nodeCatalog))
	for _, n := range nodeCatalog {
		rows = append(rows, []string{n.Name, n.Language, n.Description, n.Source})
	}
	_, _ = fmt.Fprintln(stdout, ui.Table(
		fmt.Sprintf("Installable agent nodes (%d)", len(rows)),
		[]string{"NODE", "LANG", "DESCRIPTION", "SOURCE"},
		rows,
	))
	_, _ = fmt.Fprintln(stdout)
	_, _ = fmt.Fprintln(stdout, ui.Muted("Install one with:  af install <source>"))
}

File: desktop/src/shared/catalog.ts
import type { CatalogEntry } from './types'

// Curated list of installable agent nodes, shown in the app's Install view.
//
// This is deliberately a hard-coded list maintained by hand: entries are
// vetted, and the app refuses to install any source that is not in it (the
// renderer only ever passes a catalog *name* over IPC, never a raw source).
// When the marketplace/registry search lands, this file is the seam to
// replace with a remote catalog fetch.
//
// What qualifies: an Agent-Field org repo is installable iff it has an
// `agentfield-package.yaml` manifest — at the repo root, or in a
// subdirectory addressed with the `//<subdir>` source selector (how the Go
// ports living beside their Python originals are installed). When adding an
// entry, `name` MUST equal the manifest's `name:` (the registry key after
// install — how the app detects installed state), which is often NOT the
// repo name (SWE-AF → swe-planner).
export const CATALOG: CatalogEntry[] = [
  {
    name: 'swe-planner',
    description:
      'Software factory — turn any issue into a production-ready pull request, end to end',
    source: 'https://github.com/Agent-Field/SWE-AF',
    language: 'python'
  },
  {
    name: 'swe-planner-go',
    description:
      'Software factory — turn any issue into a production-ready pull request, end to end',
    source: 'https://github.com/Agent-Field/SWE-AF//go',
    language: 'go'
  },
  {
    name: 'pr-af',
    description: 'Code review — deep, evidence-backed review of any GitHub pull request',
    source: 'https://github.com/Agent-Field/pr-af',
    language: 'python'
  },
  {
    name: 'pr-af-go',
    description: 'Code review — deep, evidence-backed review of any GitHub pull request',
    source: 'https://github.com/Agent-Field/pr-af//go',
    language: 'go'
  },
  {
    name: 'sec-af',
    description:
      'Security auditor — find vulnerabilities in your codebase and prove which ones are exploitable',
    source: 'https://github.com/Agent-Field/sec-af',
    language: 'python'
  },
  {
    name: 'cloudsecurity-af',
    description:
      'Cloud security — map real attack paths across your AWS, GCP, and Azure accounts',
    source: 'https://github.com/Agent-Field/cloudsecurity-af',
    language: 'python'
  }
]

/** Look up a catalog entry by name. Returns undefined for unknown names. */
export function catalogEntry(name: string): CatalogEntry | undefined {
  return CATALOG.find((entry) => entry.name === name)
}
